cirq/ops/phased_iswap_gate.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""ISWAPPowGate conjugated by tensor product Rz(phi) and Rz(-phi)."""
+
+from typing import List, Union, Sequence, Tuple, Optional
+
+import numpy as np
+import sympy
+
+import cirq
+from cirq import value, protocols
+from cirq._compat import proper_repr
+from cirq.ops import gate_features, raw_types, op_tree, common_gates
+from cirq.value import type_alias
+
+
+@value.value_equality(manual_cls=True)
+class PhasedISwapPowGate(common_gates.ISwapPowGate, gate_features.TwoQubitGate):
```

Comment:
I am worried about this parent-child relationship. PhasedISwapPowGate cannot actually be substituted into all places that expect an ISwapPowGate. For example, I could have code that says "if it's an iswap then it commutes with XX", but this is actually not true of a phased iswap pow gate.

Reply:
I considered and implemented two other approaches:
 * composition: putting an ISwapPowGate into `self._iswap`,
 * no direct dependency on ISwapPowGate.

In all cases, I'd like PhasedISwapPowGate to be an EigenGate since it gets us lots of things for free and makes sense conceptually (eigendecomposition is easy(*) for this gate). However, a consequence of this is that we have to store the exponent in `self._exponent`. So when you go with composition then you duplicate it between `self._iswap._exponent` and `self._exponent`.

When you remove direct dependency on ISwapPowGate, then you either end up duplicating its eigen-projectors or you end up accessing private method via `ISwapPowGate._eigen_components`.

WDYT about making `EigenGate._eigen_components` public? I think this would be the most elegant solution: doesn't break the is-a relationship (as the code in this PR does), doesn't duplicate anything and it does in fact make sense that EigenGates expose their eigendecomposition publically in some form.

----
(*) In general, if you know eigendecomposition of A then you can easily derive eigendecomposition of BAB^-1. Since conjugation is a similarity transform, eigenvalues don't change. If P is a projector on an eigenspace of A, then BPB^-1 is a projector on the corresponding eigenspace of BAB^-1.

Reply:
This still wouldn't address the problem that a phased iswap pow gate *isn't an iswap gate*. It is outside of the reference class established by the parent. It violates the Liskov substitution principle. I cannot safely pass this child class into any method that would be expecting the parent class. In my opinion this is a very serious problem.

Reply:
I agree. Done.